File: ProblemSolving/Python/Implementation/counting_valleys.py
import math
import os
import random
import re
import sys


def countingValleys(n, s):
    count = 0
    height = 0
    old_height = 0

    for i in s:
        height += 1 if i == 'U' else -1
        if height == -1 and old_height == 0:
            count += 1
        old_height = height
    return count 


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input())

    s = input()

    result = countingValleys(n, s)

    fptr.write(str(result) + '\n')

    fptr.close()

File: ProblemSolving/Go/Implementation/counting_valleys.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

func countingValleys(steps int32, path string) int32 {
	var count int32
	var height int
	for _, v := range path {
		if height == 0 && v == 'D' {
			count++
		}
		if v == 'U' {
			height++
		} else if v == 'D' {
			height--
		}
	}
	return count
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16 * 1024 * 1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16 * 1024 * 1024)

	stepsTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)
	steps := int32(stepsTemp)

	path := readLine(reader)

	result := countingValleys(steps, path)

	fmt.Fprintf(writer, "%d\n", result)

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
